src/Gaufrette/Adapter/GoogleCloudClientStorage.php

```diff
@@ -0,0 +1,340 @@
+<?php
+namespace Gaufrette\Adapter;
+
+use Gaufrette\Adapter;
+use Gaufrette\Adapter\MetadataSupporter;
+use Gaufrette\Adapter\ResourcesSupporter;
+use Gaufrette\Adapter\ListKeysAware;
+
+/**
+ * Google Cloud Storage adapter using the Google Cloud Client Library for PHP
+ * http://googlecloudplatform.github.io/google-cloud-php/
+ *
+ * @package Gaufrette
+ * @author  Lech Buszczynski <lecho@phatcat.eu>
+ */
+class GoogleCloudClientStorage implements Adapter, MetadataSupporter, ResourcesSupporter, ListKeysAware
+{
+    protected $storageClient;
+    protected $bucket;
+    protected $bucketValidated;
+    protected $options      = array();
+    protected $metadata     = array();
+    protected $resources    = array();
+
+    /**
+     * @param Google\Cloud\Storage\StorageClient    $service    Authenticated storage client class
+     * @param string                                $bucketName Name of the bucket
+     * @param array                                 $options    Options are: "directory" and "acl" (see https://cloud.google.com/storage/docs/access-control/lists)
+     */
+    public function __construct(\Google\Cloud\Storage\StorageClient $storageClient, $bucketName, $options = array())
+    {
+        $this->storageClient = $storageClient;
+        $this->setBucket($bucketName);
+        $this->options = array_replace_recursive(
+            array(
+                'directory' => '',
+                'acl'       => array()
+            ),
+            $options
+        );
+    }
+    
+    /**
+     * Get adapter options
+     * 
+     * @return  array
+     */
+    public function getOptions()
+    {
+        return $this->options;
+    }
+
+    /**
+     * Set adapter options
+     * 
+     * @param   array   $options
+     */
+    public function setOptions($options)
+    {
+        $this->options = array_replace($this->options, $options);
+    }
+    
+    protected function computePath($key)
+    {
+        if (strlen($this->options['directory']))
+        {
+            if (strcmp(substr($this->options['directory'], -1), '/') == 0)
+            {
+                return $this->options['directory'].$key;
+            }
+            return $this->options['directory'].'/'.$key;
+        }
+        return $key;
```

Comment:
Is it possible to use `/` as a key? If that's possible and you if fix previous comment, you could factor `computePath` into a one-line method :)